dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java

```diff
@@ -0,0 +1,558 @@
+import alluxio.conf.PropertyKey;
+import alluxio.exception.runtime.AlluxioRuntimeException;
+import alluxio.resource.LockResource;
+import alluxio.retry.ExponentialBackoffRetry;
+import alluxio.retry.RetryUtils;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+import com.google.common.io.Closer;
+import io.etcd.jetcd.ByteSequence;
+import io.etcd.jetcd.Client;
+import io.etcd.jetcd.KeyValue;
+import io.etcd.jetcd.Watch;
+import io.etcd.jetcd.kv.GetResponse;
+import io.etcd.jetcd.kv.PutResponse;
+import io.etcd.jetcd.lease.LeaseGrantResponse;
+import io.etcd.jetcd.lease.LeaseRevokeResponse;
+import io.etcd.jetcd.lease.LeaseTimeToLiveResponse;
+import io.etcd.jetcd.options.DeleteOption;
+import io.etcd.jetcd.options.GetOption;
+import io.etcd.jetcd.options.LeaseOption;
+import io.etcd.jetcd.options.WatchOption;
+import io.etcd.jetcd.watch.WatchEvent;
+import io.etcd.jetcd.watch.WatchResponse;
+import io.netty.util.internal.StringUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * Wrapper class around jetcd client to achieve utilities API to talk with ETCD.
+ */
+public class AlluxioEtcdClient implements Closeable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioEtcdClient.class);
+  private static final Lock INSTANCE_LOCK = new ReentrantLock();
+  @GuardedBy("INSTANCE_LOCK")
+  private static final AtomicReference<AlluxioEtcdClient> ALLUXIO_ETCD_CLIENT
+      = new AtomicReference<>();
+  private final AtomicBoolean mConnected = new AtomicBoolean(false);
+  private Client mClient;
+  public final ServiceDiscoveryRecipe mServiceDiscovery;
+  public String[] mEndpoints;
+  private final Closer mCloser = Closer.create();
+  // only watch for children change(add/remove) for given parent path
+  private final ConcurrentHashMap<String, Watch.Watcher> mRegisteredWatchers =
+      new ConcurrentHashMap<>();
+
+  /**
+   * CTOR for AlluxioEtcdClient.
+   * @param conf
+   */
+  public AlluxioEtcdClient(AlluxioConfiguration conf) {
+    String clusterName = conf.getString(PropertyKey.ALLUXIO_CLUSTER_NAME);
+    List<String> endpointsList = conf.getList(PropertyKey.ETCD_ENDPOINTS);
+    mEndpoints = endpointsList.toArray(new String[endpointsList.size()]);
```

Comment:
nit
```suggestion
    mEndpoints = endpointsList.toArray(new String[0]);
```
https://stackoverflow.com/questions/174093/toarraynew-myclass0-or-toarraynew-myclassmylist-size

Reply:
done